### PR TITLE
BSP_CAN for simulator v2

### PR DESCRIPTION
**Pull request by connorl309**
# Updates
This PR adds simulator functionality for the CAN service.

# Quality Assurance Checklist 
To make reviews more efficient, please make sure the software feature meets the following standards and check everything off that meets the quality check. Once everything has been checked, the assigned reviewers will begin the review process. _Edit this description to check off the list._

There are exceptions with all guidelines. As long as your decisions are justified, then you are good! Contact the reviewers or the leads about any exceptions. 

## Requirements 
- [x] Followed Coding Style Guide 
- [x] Code Build checks pass 
- [x] No merge conflicts 
- [x] Software feature has test associated with it. 
- [x] Software feature has documentation for it updated.
- [x] Test provides useful information and uses relevant data to accurately represent BPS 
    - NOTE: If test file already exists, use that one 
- [x] 2 Members of the team have already reviewed my PR 
    - NOTE: Approver will not look at PR if 2 other members haven't reviewed first 
- [x] Did you Have Fun?

## Things to Consider 
- [x] Even if the above are checked, is this the best way of writing my code? 
    - It's possible to write code that works, but are there ways to make code more efficient?

### CODE REVIEW

File: BSP/Simulator_V2/Src/BSP_CAN.c [hunk 1]
```diff
@@ -1,25 +1,35 @@
 /* Copyright (c) 2022 UT Longhorn Racing Solar */
 
 #include "BSP_CAN.h"
-#include "simulator_conf.h"
+#include "Simulator.h"
```
Comment (manthanand) [inline BSP/Simulator_V2/Src/BSP_CAN.c:4]: Write a test file for this to see if it actually works.

File: BSP/Simulator_V2/Src/BSP_CAN.c [hunk 2]
```diff
@@ -1,25 +1,35 @@
...
 #include <stdint.h>
 #include <stdlib.h>
 #include <sys/file.h>
 #include <unistd.h>
 
+#define to64l(arr) (((uint64_t)(((uint8_t *)(arr))[0]) <<  0)+\
```
Comment (manthanand) [inline BSP/Simulator_V2/Src/BSP_CAN.c:10]: Do you have to typecast every line or just the first one?

File: BSP/Simulator_V2/Src/BSP_CAN.c [hunk 3]
```diff
@@ -1,25 +1,35 @@
...
+                    ((uint64_t)(((uint8_t *)(arr))[1]) <<  8)+\
+                    ((uint64_t)(((uint8_t *)(arr))[2]) << 16)+\
+                    ((uint64_t)(((uint8_t *)(arr))[3]) << 24)+\
+                    ((uint64_t)(((uint8_t *)(arr))[4]) << 32)+\
+                    ((uint64_t)(((uint8_t *)(arr))[5]) << 40)+\
+                    ((uint64_t)(((uint8_t *)(arr))[6]) << 48)+\
+                    ((uint64_t)(((uint8_t *)(arr))[7]) << 56))
```
Comment (manthanand) [inline BSP/Simulator_V2/Src/BSP_CAN.c:17]: Tbh, I don't think you need to send data like this. You don't need to mimic CAN to the basics since this just means when we log the data, we will have to do some extra parsing to reverse it. Assume that the CAN protocol works correctly and just send without reversing.

File: BSP/Simulator_V2/Src/BSP_CAN.c [hunk 4]
```diff
@@ -85,6 +104,9 @@ ErrorStatus BSP_CAN_Write(uint32_t id, uint8_t data[8], uint8_t length) {
  * @return  ERROR if nothing was received so ignore id and data that was received. SUCCESS indicates data was received and stored.
  */
 ErrorStatus BSP_CAN_Read(uint32_t *id, uint8_t *data) {
+    char buffer[50];
+    sprintf(buffer, "Read CAN message of ID 0x%X\n", *id);
```
Comment (manthanand) [inline BSP/Simulator_V2/Src/BSP_CAN.c:108]: Why do you print the ID but not the data?

File: BSP/Simulator_V2/Src/Simulator.c [hunk 5]
```diff
@@ -19,6 +19,8 @@
 #include <stdbool.h>
 #include "cJSON.h" // for json parser
 
+#include "BSP_CAN.h" // CAN testing
```
Comment (manthanand) [inline BSP/Simulator_V2/Src/Simulator.c:22]: Is this required here?

File: BSP/Simulator_V2/Src/BSP_CAN.c [hunk 6]
```diff
@@ -85,6 +104,9 @@ ErrorStatus BSP_CAN_Write(uint32_t id, uint8_t data[8], uint8_t length) {
  * @return  ERROR if nothing was received so ignore id and data that was received. SUCCESS indicates data was received and stored.
  */
 ErrorStatus BSP_CAN_Read(uint32_t *id, uint8_t *data) {
+    char buffer[50];
```
[the remark this answers is not in the document]
Comment (ClarkPoon) [reply BSP/Simulator_V2/Src/BSP_CAN.c:107]: It looks like it's for sprintf, not for holding the raw CAN data

File: BSP/Simulator_V2/Src/BSP_CAN.c [hunk 7]
```diff
@@ -85,6 +104,9 @@ ErrorStatus BSP_CAN_Write(uint32_t id, uint8_t data[8], uint8_t length) {
...
+    sprintf(buffer, "Read CAN message of ID 0x%X\n", *id);
+    //gTxEnd();
```
Comment (ClarkPoon) [inline BSP/Simulator_V2/Src/BSP_CAN.c:109]: Why is this commented out?

File: BSP/Simulator_V2/Src/BSP_CAN.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (connorl309) [reply BSP/Simulator_V2/Src/BSP_CAN.c:107]: Clark is right, it's a char buffer for the entire message to print, not just the input. Regardless, CAN_READ doesn't have much functionality for some reason (Clark knows more about this).

File: BSP/Simulator_V2/Src/BSP_CAN.c (hunk 3 above)
Comment (connorl309) [reply BSP/Simulator_V2/Src/BSP_CAN.c:17]: Ok, will change this

File: BSP/Simulator_V2/Src/BSP_CAN.c (hunk 1 above)
Comment (connorl309) [reply BSP/Simulator_V2/Src/BSP_CAN.c:4]: Will do

File: BSP/Simulator_V2/Src/BSP_CAN.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (manthanand) [reply BSP/Simulator_V2/Src/BSP_CAN.c:106]: Yea, but we plan on using CAN_READ later on so we will have to get it working sometime.

File: BSP/STM32F413/Src/BSP_CAN.c [hunk 8]
```diff
@@ -241,4 +241,4 @@ void CAN1_TX_IRQHandler(void) {
     #ifdef RTOS
     OSIntExit();      // Signal to uC/OS
     #endif
-}
+}
```
Comment (manthanand) [inline BSP/STM32F413/Src/BSP_CAN.c:244]: Don't think you needed to change this file.

File: BSP/Simulator_V2/Src/BSP_CAN.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (connorl309) [reply BSP/Simulator_V2/Src/BSP_CAN.c:106]: We'd have to use the FIFO and actually write messages into it I think, in that case. Also I have some things to say about the file structure for the sim later at today's meeting

File: BSP/STM32F413/Src/BSP_CAN.c (hunk 8 above)
Comment (connorl309) [reply BSP/STM32F413/Src/BSP_CAN.c:244]: Yeah was just a typo. There weren't any content modifications made.

File: BSP/Simulator_V2/Src/Simulator.c (hunk 5 above)
Comment (connorl309) [reply BSP/Simulator_V2/Src/Simulator.c:23]: When I was doing testing directly in Sim init yes. I'll remove it once I figure out how to do a proper test file for the CAN.

File: BSP/Simulator_V2/Src/BSP_CAN.c (hunk 4 above)
Comment (connorl309) [reply BSP/Simulator_V2/Src/BSP_CAN.c:107]: See CAN_READ comments - would need to get actual CAN messages to be "transmitted" and then read into those pointers. My current code doesn't really do that because there's nowhere to grab a message from unless I use the custom FIFO, I think

File: BSP/Simulator_V2/Src/BSP_CAN.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (ClarkPoon) [reply BSP/Simulator_V2/Src/BSP_CAN.c:106]: We'll have to add support to CAN to the JSON if we want to read it. I don't think it will be too bad, but it will be more complicated than just adding a field (otherwise having a lot of CAN messages will force you to have a ton of simulator states).

It wasn't a priority for me since it doesn't work on the hardware

**Comment by connorl309** [comment]
CAN Read/Write implementation from the simulator is done, I believe. I want to work more on using the more abstract functions in cJSON for the rest of the sim but can't work on that much today/tomorrow.

File: Tests/Test_SimulatorInit.c [hunk 9]
```diff
@@ -1,6 +1,10 @@
 #include "Simulator.h"
+#include "BSP_CAN.h"
 
 int main(void) {
     Simulator_init();
+    BSP_CAN_Init(NULL, NULL, false);
+    uint8_t data[8] = "NICECOCK";
```
Comment (manthanand) [inline Tests/Test_SimulatorInit.c:7]: Let's consider a more elegant test case. Also, a test for CAN should be made in its own file, with a header that explains how the test should be conducted.

Comment (connorl309) [reply Tests/Test_SimulatorInit.c:7]: whoops thats my bad forgot about that will change and look into making an actual test

Comment (connorl309) [reply Tests/Test_SimulatorInit.c]: > Let's consider a more elegant test case. Also, a test for CAN should be made in its own file, with a header that explains how the test should be conducted.

The one thing with doing tests in a .c file is that all our data is being read from the JSON files, so I don't really see how we can specify which JSON file to use.

Comment (connorl309) [reply Tests/Test_SimulatorInit.c]: nvm working on it

File: BSP/Simulator_V2/Src/BSP_CAN.c [hunk 10]
```diff
@@ -11,15 +12,16 @@
...
 /**
  * @brief   Initializes the CAN module that communicates with the rest of the electrical system.
  * @param   None
  * @return  None
  */
-
-
-
 void BSP_CAN_Init(callback_t rxEvent, callback_t txEnd, bool loopback) {
 #if 0   // TODO: replace with interrupt-capable code
```
Comment (manthanand) [inline BSP/Simulator_V2/Src/BSP_CAN.c:26]: remove all the code that was taken from the old simulator

File: BSP/Simulator_V2/Data/cantest.json [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+                "id": 0,
+                "data": "DEADBEEF",
+                "len": 8
+            },
+            "msg2": {
+                "read_write": "write",
+                "id": 100,
+                "data": "LHRSOLAR",
+                "len": 8
+            }
```
[the remark this answers is not in the document]
Comment (connorl309) [reply BSP/Simulator_V2/Data/cantest.json:27]: Will look into this once simulator rewrite gets merged

**Review by MichaelMohn — APPROVED** [review]
Test code seems right.

**Comment by ClarkPoon** [comment]
Fix the merge conflicts

File: Docs/source/Simulator/Overview.rst [hunk 12]
```diff
@@ -1,49 +0,0 @@
-********
```
Comment (manthanand) [inline Docs/source/Simulator/Overview.rst:1]: Is there a reason this was deleted?

File: Docs/source/index.rst [hunk 13]
```diff
@@ -38,13 +38,6 @@ Welcome to the BeVolt BPS documentation!
    Hardware/Components
    Hardware/battery-compartment.rst
 
-.. toctree::
```
Comment (manthanand) [inline Docs/source/index.rst:41]: Why was this deleted
